coachmark/src/main/java/com/pseudoankit/coachmark/overlay/OverlayLayout.kt

```diff
@@ -0,0 +1,147 @@
+package com.pseudoankit.coachmark.overlay
+
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.UiComposable
+import androidx.compose.ui.layout.Layout
+import androidx.compose.ui.layout.Measurable
+import androidx.compose.ui.layout.Placeable
+import androidx.compose.ui.layout.layoutId
+import androidx.compose.ui.unit.Constraints
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import com.pseudoankit.coachmark.model.ToolTipPlacement
+import com.pseudoankit.coachmark.model.TooltipConfig
+
+/** Containing composable must use these values for layoutId on current and previous tooltip. */
+public enum class OverlayChildLayoutId { CURRENT, PREVIOUS }
+
+/** Proposed minimum useful tooltip width; this can be adjusted if needed */
+private const val TOOLTIP_MAX_WIDTH_OVERRIDE_PX = 30
+
+/**
+ * Extracted from DimOverlayEffect, in case another overlay effect needs it in the future.
+ *
+ * @param configCurrent of coach mark highlight; null suits case where there's no "current" tooltip in content
+ * @param configPrevious of coach mark highlight; null suits case where there's no "previous" tooltip in content
+ * @param gapTooltipScreen min distance between tooltip and edge of screen
+ */
+@Composable
+public fun OverlayLayout(
+    content: @Composable @UiComposable () -> Unit,
+    configCurrent: TooltipConfig?,
+    configPrevious: TooltipConfig?,
+    modifier: Modifier = Modifier,
+    gapTooltipScreen: Dp = 8.dp,
+) {
+    Layout(content, modifier) { measurables, constraints ->
+
+        // child count < 2 occurs on first and last coach mark
+        require(measurables.size <= 2) { "OverlayLayout cannot have more than two children" }
+
+        val gapTooltipScreenPx = gapTooltipScreen.roundToPx()
+
+        // measure children
+        val placeableCurrent = measure(configCurrent, OverlayChildLayoutId.CURRENT, measurables, constraints, gapTooltipScreenPx)
+        val placeablePrevious = measure(configPrevious, OverlayChildLayoutId.PREVIOUS, measurables, constraints, gapTooltipScreenPx)
+
+        // place children
+        layout(constraints.maxWidth, constraints.maxHeight) {
+            place(placeableCurrent, configCurrent)
+            place(placeablePrevious, configPrevious)
+        }
+    }
+}
+
+
+/**
+ * @return null if tooltipConfig is null
+ */
+private fun measure(
+    tooltipConfig: TooltipConfig?,
+    layoutId: OverlayChildLayoutId,
+    measurables: List<Measurable>,
+    constraintsParent: Constraints,
+    gapTooltipScreenPx: Int,
+): Placeable? {
+    if (tooltipConfig == null) return null
+
+    // constrain max width to prevent tooltip running off screen
+    var maxWidth = when (tooltipConfig.toolTipPlacement) {
+        ToolTipPlacement.Start -> {
+            constraintsParent.maxWidth - gapTooltipScreenPx - (constraintsParent.maxWidth - tooltipConfig.layout.startX.toInt())
```

Comment:
do we need `constraintsParent.maxWidth` here, as both values will be cancelled right?

Reply:
since we are trying to calculate tooltip max width shouldn't it be
constraintsParent.maxWidth - gapTooltipScreenPx - tooltipConfig.layout. endX.toInt() (since this is for start) ? or can you please explain the calcuations here

Reply:
I'll at least double-check it and add a comment